diff --git a/cmd/kubeadm/app/cmd/join.go b/cmd/kubeadm/app/cmd/join.go
index fe970e0860c..9478c2b8f8a 100644
--- a/cmd/kubeadm/app/cmd/join.go
+++ b/cmd/kubeadm/app/cmd/join.go
@@ -140,6 +140,8 @@ type joinOptions struct {
 	patchesDir            string
 	dryRun                bool
 	skipCRIDetect         bool
+	kubeconfigDir         string
+	kubeconfigPath         string
 }
 
 // compile-time assert that the local data object satisfies the phases data interface.
@@ -157,6 +159,8 @@ type joinData struct {
 	patchesDir            string
 	dryRun                bool
 	dryRunDir             string
+	kubeconfigDir         string
+	kubeconfigPath         string
 }
 
 // newCmdJoin returns "kubeadm join" command.
@@ -219,6 +223,13 @@ func newCmdJoin(out io.Writer, joinOptions *joinOptions) *cobra.Command {
 	addJoinConfigFlags(cmd.Flags(), joinOptions.externalcfg)
 	addJoinOtherFlags(cmd.Flags(), joinOptions)
 
+	// defines additional flags that are not used by the join command but that could be eventually used
+	// by the sub-commands automatically generated for phases
+	joinRunner.SetAdditionalFlags(func(flags *flag.FlagSet) {
+		options.AddKubeConfigDirFlag(flags, &joinOptions.kubeconfigDir)
+		options.AddKubeConfigFlag(flags, &joinOptions.kubeconfigPath)
+	})
+
 	joinRunner.AppendPhase(phases.NewPreflightPhase())
 	joinRunner.AppendPhase(phases.NewControlPlanePreparePhase())
 	joinRunner.AppendPhase(phases.NewCheckEtcdPhase())
@@ -236,7 +247,11 @@ func newCmdJoin(out io.Writer, joinOptions *joinOptions) *cobra.Command {
 			// assume that the command execution does not depend on CRISocket when --cri-socket flag is not set
 			joinOptions.skipCRIDetect = true
 		}
-		data, err := newJoinData(cmd, args, joinOptions, out, kubeadmconstants.GetAdminKubeConfigPath())
+    		configPath := kubeadmconstants.GetAdminKubeConfigPath()
+		if configFlag := cmd.Flags().Lookup(options.KubeconfigPath); configFlag != nil {
+    		    configPath = configFlag.Value.String()
+		}
+		data, err := newJoinData(cmd, args, joinOptions, out, configPath)
 		if err != nil {
 			return nil, err
 		}
@@ -340,6 +355,8 @@ func newJoinOptions() *joinOptions {
 
 	return &joinOptions{
 		externalcfg: externalcfg,
+		kubeconfigDir: kubeadmconstants.KubernetesDir,
+		kubeconfigPath: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName),
 	}
 }
 
@@ -481,6 +498,8 @@ func newJoinData(cmd *cobra.Command, args []string, opt *joinOptions, out io.Wri
 		outputWriter:          out,
 		patchesDir:            opt.patchesDir,
 		dryRunDir:             dryRunDir,
+		kubeconfigDir:         opt.kubeconfigDir,
+		kubeconfigPath:         opt.kubeconfigPath,
 	}, nil
 }
 
@@ -507,7 +526,14 @@ func (j *joinData) KubeConfigDir() string {
 	if j.dryRun {
 		return j.dryRunDir
 	}
-	return kubeadmconstants.KubernetesDir
+	return j.kubeconfigDir
+}
+
+func (j *joinData) KubeConfigPath() string {
+	if j.dryRun {
+		j.kubeconfigPath = filepath.Join(j.dryRunDir, kubeadmconstants.AdminKubeConfigFileName)
+	}
+	return j.kubeconfigPath
 }
 
 // KubeletDir returns the kubelet configuration directory or the temporary directory if DryRun is true.
@@ -580,7 +606,7 @@ func (j *joinData) InitCfg() (*kubeadmapi.InitConfiguration, error) {
 
 // Client returns the Client for accessing the cluster with the identity defined in admin.conf.
 func (j *joinData) Client() (clientset.Interface, error) {
-	pathAdmin := filepath.Join(j.KubeConfigDir(), kubeadmconstants.AdminKubeConfigFileName)
+	pathAdmin := j.KubeConfigPath()
 
 	if j.dryRun {
 		dryRun := apiclient.NewDryRun()
@@ -632,7 +658,7 @@ func (j *joinData) Client() (clientset.Interface, error) {
 // It uses the admin.conf as the base, but modifies it to point at the local API server instead
 // of the control plane endpoint.
 func (j *joinData) WaitControlPlaneClient() (clientset.Interface, error) {
-	pathAdmin := filepath.Join(j.KubeConfigDir(), kubeadmconstants.AdminKubeConfigFileName)
+	pathAdmin := j.KubeConfigPath()
 	config, err := clientcmd.LoadFromFile(pathAdmin)
 	if err != nil {
 		return nil, err
diff --git a/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go b/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
index 242d289402a..32fec5c2c3d 100644
--- a/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
+++ b/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
@@ -47,7 +47,10 @@ func getControlPlaneJoinPhaseFlags(name string) []string {
 		options.NodeName,
 	}
 	if name == "etcd" || name == "all" {
-		flags = append(flags, options.Patches)
+		flags = append(flags,
+    		        options.Patches,
+    		        options.KubeconfigPath,
+		)
 	}
 	if name != "mark-control-plane" {
 		flags = append(flags, options.APIServerAdvertiseAddress)
diff --git a/cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go b/cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go
index ea01f3db686..5ee0740e3ba 100644
--- a/cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go
+++ b/cmd/kubeadm/app/cmd/phases/join/controlplaneprepare.go
@@ -113,6 +113,7 @@ func getControlPlanePreparePhaseFlags(name string) []string {
 	case "kubeconfig":
 		flags = []string{
 			options.CfgPath,
+			options.KubeconfigDir,
 			options.ControlPlane,
 			options.FileDiscovery,
 			options.TokenDiscovery,
diff --git a/cmd/kubeadm/app/cmd/phases/join/data.go b/cmd/kubeadm/app/cmd/phases/join/data.go
index 197d018b75d..07be944093f 100644
--- a/cmd/kubeadm/app/cmd/phases/join/data.go
+++ b/cmd/kubeadm/app/cmd/phases/join/data.go
@@ -41,6 +41,7 @@ type JoinData interface {
 	PatchesDir() string
 	DryRun() bool
 	KubeConfigDir() string
+	KubeConfigPath() string
 	KubeletDir() string
 	ManifestDir() string
 	CertificateWriteDir() string
@@ -63,6 +64,7 @@ func (j *testJoinData) OutputWriter() io.Writer                              { r
 func (j *testJoinData) PatchesDir() string                                   { return "" }
 func (j *testJoinData) DryRun() bool                                         { return false }
 func (j *testJoinData) KubeConfigDir() string                                { return "" }
+func (j *testJoinData) KubeConfigPath() string                               { return "" }
 func (j *testJoinData) KubeletDir() string                                   { return "" }
 func (j *testJoinData) ManifestDir() string                                  { return "" }
 func (j *testJoinData) CertificateWriteDir() string                          { return "" }
diff --git a/cmd/kubeadm/app/cmd/phases/join/kubelet.go b/cmd/kubeadm/app/cmd/phases/join/kubelet.go
index fec28a526bc..5ca7e2eeeb6 100644
--- a/cmd/kubeadm/app/cmd/phases/join/kubelet.go
+++ b/cmd/kubeadm/app/cmd/phases/join/kubelet.go
@@ -65,6 +65,7 @@ func NewKubeletStartPhase() workflow.Phase {
 			options.TokenStr,
 			options.Patches,
 			options.DryRun,
+			options.KubeconfigDir,
 		},
 	}
 }
@@ -79,6 +80,8 @@ func NewKubeletWaitBootstrapPhase() workflow.Phase {
 			options.CfgPath,
 			options.NodeCRISocket,
 			options.DryRun,
+			options.KubeconfigPath,
+			options.KubeconfigDir,
 		},
 	}
 }
@@ -281,7 +284,8 @@ func runKubeletWaitBootstrapPhase(c workflow.RunData) (returnErr error) {
 		return errors.Wrap(err, "failed while waiting for the kubelet to start")
 	}
 
-	if err := waitForTLSBootstrappedClient(cfg.Timeouts.TLSBootstrap.Duration); err != nil {
+	kubeConfigFile := filepath.Join(data.KubeConfigDir(), kubeadmconstants.KubeletKubeConfigFileName)
+	if err := waitForTLSBootstrappedClient(cfg.Timeouts.TLSBootstrap.Duration, kubeConfigFile); err != nil {
 		apiclient.PrintKubeletErrorHelpScreen(data.OutputWriter())
 		return errors.Wrap(err, "failed while waiting for TLS bootstrap")
 	}
@@ -290,7 +294,7 @@ func runKubeletWaitBootstrapPhase(c workflow.RunData) (returnErr error) {
 }
 
 // waitForTLSBootstrappedClient waits for the /etc/kubernetes/kubelet.conf file to be available
-func waitForTLSBootstrappedClient(timeout time.Duration) error {
+func waitForTLSBootstrappedClient(timeout time.Duration, filename string) error {
 	fmt.Println("[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap")
 
 	// Loop on every falsy return. Return with an error if raised. Exit successfully if true is returned.
@@ -300,7 +304,7 @@ func waitForTLSBootstrappedClient(timeout time.Duration) error {
 			// Check that we can create a client set out of the kubelet kubeconfig. This ensures not
 			// only that the kubeconfig file exists, but that other files required by it also exist (like
 			// client certificate and key)
-			_, err := kubeconfigutil.ClientSetFromFile(kubeadmconstants.GetKubeletKubeConfigPath())
+			_, err := kubeconfigutil.ClientSetFromFile(filename)
 			return (err == nil), nil
 		})
 }
