diff -ur rustc-1.83.0-src/src/bootstrap/src/core/builder.rs rustc-patched/src/bootstrap/src/core/builder.rs
--- rustc-1.83.0-src/src/bootstrap/src/core/builder.rs	2024-11-26 02:12:09.000000000 -0800
+++ rustc-patched/src/bootstrap/src/core/builder.rs	2024-12-29 22:17:02.905101938 -0800
@@ -22,7 +22,7 @@
 use crate::utils::cache::Cache;
 use crate::utils::exec::{BootstrapCommand, command};
 use crate::utils::helpers::{
-    self, LldThreads, add_dylib_path, add_link_lib_path, check_cfg_arg, exe, libdir, linker_args,
+    self, LldThreads, add_dylib_path, check_cfg_arg, exe, libdir, linker_args,
     linker_flags, t,
 };
 use crate::{
@@ -2003,7 +2003,7 @@
             if let Some(llvm_config) = self.llvm_config(target) {
                 let llvm_libdir =
                     command(llvm_config).arg("--libdir").run_capture_stdout(self).stdout();
-                add_link_lib_path(vec![llvm_libdir.trim().into()], &mut cargo);
+                rustflags.arg(&format!("-Clink-arg=-L{llvm_libdir}"));
             }
         }
 
diff -ur rustc-1.83.0-src/src/bootstrap/src/core/build_steps/test.rs rustc-patched/src/bootstrap/src/core/build_steps/test.rs
--- rustc-1.83.0-src/src/bootstrap/src/core/build_steps/test.rs	2024-11-26 02:12:09.000000000 -0800
+++ rustc-patched/src/bootstrap/src/core/build_steps/test.rs	2024-12-29 22:17:06.680135838 -0800
@@ -21,7 +21,7 @@
 use crate::core::config::flags::{Subcommand, get_completion};
 use crate::utils::exec::{BootstrapCommand, command};
 use crate::utils::helpers::{
-    self, LldThreads, add_link_lib_path, add_rustdoc_cargo_linker_args, dylib_path, dylib_path_var,
+    self, LldThreads, add_rustdoc_cargo_linker_args, dylib_path, dylib_path_var,
     linker_args, linker_flags, t, target_supports_cranelift_backend, up_to_date,
 };
 use crate::utils::render_tests::{add_flags_and_try_run_tests, try_run_tests};
@@ -1962,7 +1962,7 @@
             if !builder.config.dry_run() && suite.ends_with("fulldeps") {
                 let llvm_libdir =
                     command(&llvm_config).arg("--libdir").run_capture_stdout(builder).stdout();
-                add_link_lib_path(vec![llvm_libdir.trim().into()], &mut cmd);
+                cmd.env("RUSTFLAGS", format!("-Clink-arg=-L{llvm_libdir}"));
             }
 
             if !builder.config.dry_run() && matches!(mode, "run-make" | "coverage-run") {
diff -ur rustc-1.83.0-src/src/bootstrap/src/utils/helpers.rs rustc-patched/src/bootstrap/src/utils/helpers.rs
--- rustc-1.83.0-src/src/bootstrap/src/utils/helpers.rs	2024-11-26 02:12:09.000000000 -0800
+++ rustc-patched/src/bootstrap/src/utils/helpers.rs	2024-12-29 22:17:36.601404446 -0800
@@ -79,31 +79,6 @@
     cmd.env(dylib_path_var(), t!(env::join_paths(list)));
 }
 
-/// Adds a list of lookup paths to `cmd`'s link library lookup path.
-pub fn add_link_lib_path(path: Vec<PathBuf>, cmd: &mut BootstrapCommand) {
-    let mut list = link_lib_path();
-    for path in path {
-        list.insert(0, path);
-    }
-    cmd.env(link_lib_path_var(), t!(env::join_paths(list)));
-}
-
-/// Returns the environment variable which the link library lookup path
-/// resides in for this platform.
-fn link_lib_path_var() -> &'static str {
-    if cfg!(target_env = "msvc") { "LIB" } else { "LIBRARY_PATH" }
-}
-
-/// Parses the `link_lib_path_var()` environment variable, returning a list of
-/// paths that are members of this lookup path.
-fn link_lib_path() -> Vec<PathBuf> {
-    let var = match env::var_os(link_lib_path_var()) {
-        Some(v) => v,
-        None => return vec![],
-    };
-    env::split_paths(&var).collect()
-}
-
 pub struct TimeIt(bool, Instant);
 
 /// Returns an RAII structure that prints out how long it took to drop.
