diff --git a/etc/rc b/etc/rc
index 9d87fac8caf..ed2e6dc17c2 100644
--- a/etc/rc
+++ b/etc/rc
@@ -337,7 +337,7 @@ trap : 3	# Shouldn't be needed.
 
 export HOME=/
 export INRC=1
-export PATH=/sbin:/bin:/usr/sbin:/usr/bin
+#export PATH=/sbin:/bin:/usr/sbin:/usr/bin
 
 # /etc/myname contains my symbolic name.
 if [[ -f /etc/myname ]]; then
@@ -353,6 +353,10 @@ fi
 FUNCS_ONLY=1 . /etc/rc.d/rc.subr
 _rc_parse_conf
 
+daemon_phase() {
+	. /etc/rc.daemon/$1
+}
+
 # If executed with the 'shutdown' parameter by the halt, reboot or shutdown:
 # - update seed files
 # - execute the rc.d scripts specified by $pkg_scripts in reverse order
@@ -414,7 +418,7 @@ fi
 trap "echo 'Boot interrupted.'; exit 1" 3
 
 # Unmount all filesystems except root.
-umount -a >/dev/null 2>&1
+#umount -a >/dev/null 2>&1
 
 # Mount all filesystems except those of type NFS and VND.
 mount -a -t nonfs,vnd
@@ -483,9 +487,10 @@ mount -s /var >/dev/null 2>&1		# cannot be on NFS
 mount -s /var/log >/dev/null 2>&1	# cannot be on NFS
 mount -s /usr >/dev/null 2>&1		# if NFS, fstab must use IP address
 
-reorder_libs 2>&1 |&
+#reorder_libs 2>&1 |&
 
-start_daemon slaacd dhcpleased resolvd >/dev/null 2>&1
+#start_daemon slaacd dhcpleased resolvd >/dev/null 2>&1
+daemon_phase 10
 
 echo 'starting network'
 
@@ -495,11 +500,12 @@ ifconfig -g carp carpdemote 128
 
 sh /etc/netstart
 
-start_daemon unwind >/dev/null 2>&1
+#start_daemon unwind >/dev/null 2>&1
+daemon_phase 20
 
 random_seed
 
-wait_reorder_libs
+#wait_reorder_libs
 
 # Load pf rules and bring up pfsync interface.
 if [[ $pf != NO ]]; then
@@ -522,8 +528,9 @@ dmesg >/var/run/dmesg.boot
 make_keys
 
 echo -n 'starting early daemons:'
-start_daemon syslogd ldattach pflogd nsd unbound ntpd
-start_daemon iscsid isakmpd iked sasyncd ldapd npppd
+#start_daemon syslogd ldattach pflogd nsd unbound ntpd
+#start_daemon iscsid isakmpd iked sasyncd ldapd npppd
+daemon_phase 30
 echo '.'
 
 # Load IPsec rules.
@@ -532,11 +539,12 @@ if [[ $ipsec != NO && -f /etc/ipsec.conf ]]; then
 fi
 
 echo -n 'starting RPC daemons:'
-start_daemon portmap
-if [[ -n $(domainname) ]]; then
-	start_daemon ypldap ypserv ypbind
-fi
-start_daemon mountd nfsd lockd statd amd
+#start_daemon portmap
+#if [[ -n $(domainname) ]]; then
+#	start_daemon ypldap ypserv ypbind
+#fi
+#start_daemon mountd nfsd lockd statd amd
+daemon_phase 40
 echo '.'
 
 # Check and mount remaining file systems and enable additional swap.
@@ -626,30 +634,32 @@ echo 'preserving editor files.'; /usr/libexec/vi.recover
 run_upgrade_script sysmerge
 
 echo -n 'starting network daemons:'
-start_daemon ldomd sshd snmpd ldpd ripd ospfd ospf6d bgpd ifstated
-start_daemon relayd dhcpd dhcrelay mrouted dvmrpd radiusd eigrpd route6d
-start_daemon rad hostapd lpd smtpd slowcgi bgplgd httpd ftpd
-start_daemon ftpproxy ftpproxy6 tftpd tftpproxy identd inetd rarpd bootparamd
-start_daemon rbootd mopd vmd spamd spamlogd sndiod
+#start_daemon ldomd sshd snmpd ldpd ripd ospfd ospf6d bgpd ifstated
+#start_daemon relayd dhcpd dhcrelay mrouted dvmrpd radiusd eigrpd route6d
+#start_daemon rad hostapd lpd smtpd slowcgi bgplgd httpd ftpd
+#start_daemon ftpproxy ftpproxy6 tftpd tftpproxy identd inetd rarpd bootparamd
+#start_daemon rbootd mopd vmd spamd spamlogd sndiod
+daemon_phase 50
 echo '.'
 
 # If rc.firsttime exists, run it just once, and make sure it is deleted.
 run_upgrade_script firsttime
 
 # Run rc.d(8) scripts from packages.
-if [[ -n $pkg_scripts ]]; then
-	echo -n 'starting package daemons:'
-	for _daemon in $pkg_scripts; do
-		if [[ -x /etc/rc.d/$_daemon ]]; then
-			start_daemon $_daemon
-		else
-			echo -n " ${_daemon}(absent)"
-		fi
-	done
-	echo '.'
-fi
-
-[[ -f /etc/rc.local ]] && sh /etc/rc.local
+#if [[ -n $pkg_scripts ]]; then
+#	echo -n 'starting package daemons:'
+#	for _daemon in $pkg_scripts; do
+#		if [[ -x /etc/rc.d/$_daemon ]]; then
+#			start_daemon $_daemon
+#		else
+#			echo -n " ${_daemon}(absent)"
+#		fi
+#	done
+#	echo '.'
+#fi
+#
+#[[ -f /etc/rc.local ]] && sh /etc/rc.local
+daemon_phase 60
 
 # Disable carp interlock.
 ifconfig -g carp -carpdemote 128
@@ -657,12 +667,13 @@ ifconfig -g carp -carpdemote 128
 mixerctl_conf
 
 echo -n 'starting local daemons:'
-start_daemon apmd sensorsd hotplugd watchdogd cron wsmoused xenodm
+#start_daemon apmd sensorsd hotplugd watchdogd cron wsmoused xenodm
+daemon_phase 70
 echo '.'
 
 # Re-link the kernel, placing the objects in a random order.
 # Replace current with relinked kernel and inform root about it.
-/usr/libexec/reorder_kernel &
+#/usr/libexec/reorder_kernel &
 
 date
 exit 0
