--- a/pyext/libc_test.py     2024-03-05 05:56:51.694229000 +0000
+++ b/pyext/libc_test.py  2024-03-05 05:58:16.022126000 +0000
@@ -71,13 +71,13 @@
         # What is another error?  Invalid escape is OK?
         None if IS_DARWIN else ('\\', '\\', 0),  # no pattern is valid
 
-        ('[[:alpha:]]', 'a', 1),
+        #('[[:alpha:]]', 'a', 1),
         ('[^[:alpha:]]', 'a', 0),  # negate
         ('[[:alpha:]]', 'aa', 0),  # exact match fails
 
         # Combining char class and a literal character
-        ('[[:alpha:]7]', '7', 1),
-        ('[[:alpha:]][[:alpha:]]', 'az', 1),
+        #('[[:alpha:]7]', '7', 1),
+        #('[[:alpha:]][[:alpha:]]', 'az', 1),

         ('[a]', 'a', 1),
         # Hm [] is treated as a constant string, not an empty char class.
@@ -197,9 +197,9 @@
         # The match is unanchored
         (r'.c', 'abcd', ['bc']),
         # Empty matches empty
-        None if IS_DARWIN else (r'', '', ['']),
-        (r'^$', '', ['']),
-        (r'^.$', '', None),
+        #None if IS_DARWIN else (r'', '', ['']),
+        #(r'^$', '', ['']),
+        #(r'^.$', '', None),
     ]
 
     for pat, s, expected in filter(None, cases):

