--- libwnck-43.0/libwnck/meson.build	2024-02-27 16:59:16.524416000 +0000
+++ libwnck-43.0/libwnck/meson.build	2024-02-27 17:04:36.368760000 +0000
@@ -118,7 +118,6 @@
   dependencies: LIBWNCK_DEPS,
   compile_args: libwnck_cflags,
   sources: headers + [enum_types[1]],
-  link_args: libwnck_ldflags,
 )
 
 libwnck_lib = shared_library(LIBNAME,
@@ -126,6 +125,7 @@
   sources: sources + a11y_sources + enum_types + resources,
   version: '@0@.@1@.@2@'.format(LIBWNCK_SOVERSION, LIBWNCK_CURRENT, LIBWNCK_REVISION),
   soversion: LIBWNCK_SOVERSION,
+  link_args: libwnck_ldflags,
   link_depends: mapfile,
   install: true,
 )
